### DISCUSSION TITLE
`Missing or invalid. request() method: eth_call`

### DISCUSSION THREAD

**Discussion by efstajas** [WalletConnect/walletconnect-monorepo#2660] [Developer Support]
Hey there,

I'm currently trying to migrate our app (github.com/radicle-dev/drips-app-v2) to WalletConnect V2, and at the same time upgrade from Web3Modal V1 to V2. Unfortunately, there are two main complications:

- Our app & internal SDK are built on Ethers, and require Providers and Signers
- We don't use React but SvelteKit, so the recommended Wagmi solution doesn't just work out of the box

Despite all that, I managed to get the new Web3Modal working almost fully in this PR: https://github.com/radicle-dev/drips-app-v2/pull/465 Because we do probably want to migrate to `viem` at some point, I went down the vanilla JS `wagmi` route, and for now just [created an Ethers provider from viem's `WalletClient` by passing its RPC `request` function to initialize an Ethers `Web3Provider` that talks to viem behind the scenes](https://github.com/radicle-dev/drips-app-v2/pull/465/files#diff-b2fb45df4a0b1421037bf0b8501ac750de5b864f4328dd451b54039e6939dfa6R177-R190).

This is obviously super dirty, and I wonder if there's a better approach. Either way, the problem is that when I set `version` to `2`, the following happens:

- Open web3Modal
- Scan the QR code with a V2-enabled wallet
- web3Modal closes
- the app throws the following exception as soon as it attempts an `eth_call`: 

```
{"details":"Missing or invalid. request() method: eth_call","shortMessage":"An unknown RPC error occurred.","name":"UnknownRpcError","version":"viem@1.0.7","message":"An unknown RPC error occurred.\n\nDetails: Missing or invalid. request() method: eth_call\nVersion: viem@1.0.7","cause":{},"code":-1}
```

What is `Missing or invalid` referring to here exactly? I've double-checked the validity of the `eth_call` itself. Of course, everything works fine when connecting to MetaMask (also via Web3Modal).

When I set `version` back to `1`, things work _better_, but also not fully. When I trigger the modal and connect a wallet via the QR code for the very first time after starting the app, the modal just closes and an exception is thrown, again as soon as it attempts an `eth_call`:

```
{"details":"Cannot request JSON-RPC method (eth_call) without provided rpc url","shortMessage":"An unknown RPC error occurred.","name":"UnknownRpcError","version":"viem@1.0.7","message":"An unknown RPC error occurred.\n\nDetails: Cannot request JSON-RPC method (eth_call) without provided rpc url\nVersion: viem@1.0.7","cause":{},"code":-1}
```

However, here the weird thing is that if I reload the app right after, WalletConnect restores the previous connection as expected, and now everything suddenly works fine. It also works fine when first connecting MetaMask, then disconnecting, and then connecting via WalletConnect. This error only occurs when connecting via WalletConnect straight after opening the app for the first time.

I've tried:
- completely nuking and reinstalling `node_modules`, even with a fresh lock file
- different wallets to scan the QR code
- different browsers
- connecting to different networks (mainnet & goerli)
- upgrading a bunch of dependencies like vite, sveltekit, typescript etc. to their latest versions

... but all to no avail.

I realize this is probably a highly specific question, and I've probably coded myself into a very awkward corner... but does anyone have any ideas?

**Comment by glitch-txs** [comment]
Hey which wallet are you testing this with?
Does this happens in labs too?
https://lab.web3modal.com/

**Reply by efstajas** [reply]
I'm using Rainbow on Android, but also tried MetaMask mobile. 

Everything seems to be working fine on the lab, so it's probably something wrong with my implementation.

**Reply by glitch-txs** [reply]
MetaMask doesn't support v2 yet, could try with Trust Wallet?

**Reply by efstajas** [reply]
Ah, makes sense then that it wouldn't work. 

Same exact situation with Trust Wallet – works fine on the lab, but not with my implementation.

(Btw, please note that I updated my post very shortly before you commented. It turned out that the `Not found` error resolved itself after I cleared my `package-lock.json`, so the odd RPC errors are the only remaining problem)

**Reply by glitch-txs** [reply]
Do you think you can create a minimal reproducible example for this? I can try help you debug

**Reply by efstajas** [reply]
I think so – based on what I've found out so far I suspect the same would happen in a plain environment when trying to convert the `viem` WalletClient into an Ethers provider. Need to wrap up for today now, but I'll try to do this first thing on Monday.

Thanks a lot for your quick help already!

**Reply by efstajas** [reply]
So after spending some more time on this yesterday, I decided to give up on Web3Modal and try some other approaches. Eventually ended up with `web3-onboard`: https://github.com/radicle-dev/drips-app-v2/pull/467 It was super straightforward to set up and I didn't have any of the problems described above, even when using WC Protocol V2.

I'm going to leave this thread open for future reference for now.

**Reply by finessevanes** [reply]
@efstajas Hey, I'm sorry you had a bad experience attempting to implement Web3Modal. I'm curious if you'd be interested in a UX interview to hear your feedback and improve docs 🙏🏽

**Reply by efstajas** [reply]
Hey @finessevanes yeah, i'd be interested for sure. My email is on my profile, feel free to reach out anytime!

**Comment by lm93547** [comment]
I have almost the identical issue. Why would `eth_call` not work on V2?

**Reply by boidushya** [reply]
@lm93547 Hey! If you're using wagmi with ethers, this issue should be resolved. I've created a [codesandbox example](https://codesandbox.io/p/sandbox/hungry-noether-ltwc5g) to help demonstrate this fix. 
TL;DR: you need to update wagmi to `0.12.16` or higher and under `signerOrProvider` option pass in the `signer` you receive from wagmi's `useSigner` hook.
If you run into any further issues with the implementation, feel free to reply to this thread

**Reply by lm93547** [reply]
I am on wagmi `1.1.1` and the `useSigner` hook does not exist. I can try the `useWalletClient` but because it uses `viem` under the hood the implementation is different to the `Ethers.JsonRpcSigner` that `ethers` contracts expects. I have also tried converting the walletClient to the signer as described here: https://wagmi.sh/react/ethers-adapters and still seeing the same errors.

**Reply by jxom** [reply]
wagmi v1 doesn't use `eth_call` via WalletConnect v2 at all. It goes via the RPC Provider you provide to `createConfig`. Can you provide a minimal reproduction?
